client/client.go

```diff
@@ -136,49 +220,74 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
 	return
 }
 
-func (client *Client) reconnect(err error) error {
-	if client.conn != nil {
-		return nil
+func (client *Client) grabReconnectState() (success bool) {
+	client.reconnectLock.Lock()
+	defer client.reconnectLock.Unlock()
+	if client.reconnectingActive { // another thread is already reconnecting to server, this thread will exit
+		return false
 	}
+	client.reconnectingActive = true // I am first to attempt reconnection
+	return true
+}
+
+// called by owner of reconnect state to tell that it has finished reconnecting
+func (client *Client) resetReconnectState() {
+	client.reconnectingActive = false
+}
 
-	// TODO I doubt this error handling is right because it looks
-	// really complicated.
+func (client *Client) reconnect(err error) error {
```

Comment:
I think some sort of atomic reconnection control makes sense, but this combined with the grabReconnectState() makes it pretty hard to reason about. Can you lay out the reconnect semantics we expect? What do we want to have happen if two reconnect() calls are invoked concurrently?

Reply:
@jacksonrnewhouse 
1. Lock is to prevent the other of the 2 threads from using the connection for actual send/receive.
2. grabReconectState() helps resolve the race between the two threads if both failed at (nearly) same time and attempting to reconnect.

Reply:
update: removed client Lock, changed read/write loop reconnect behavior to both exit and look at conn==nil as indicator that reconnection is in progress.